### ISSUE TITLE
Target(string) ctor invokes JIT if the string contains "cuda"

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#4708]
If you execute a generator binary with `target=anything-cuda`, this happens to call `Target(string)`, which calls `merge_string`, which calls `get_host_cuda_capability()` if the string contains `cuda`... which ends up calling `get_device_interface_for_device_api()`, which ends up calling `JITSharedRuntime::get`, which basically initializes all of the JIT's shared runtime code.

This is a nontrivial and unexpected cost for simply constructing a `Target` from a string. Surely we can circumvent this?

**Comment by abadams** [comment]
We need some way to sniff the host cuda capability to avoid using lowest-common-denominator features. Initializing the runtime is simplest, because that's where the code capable of loading libcuda is, and it's very likely to if you're saying host-cuda, you're about to compile cuda code for the host anyway. We could try to sniff the capability some other way, but it will at least involve loading a shared library, possibly redundantly.

When we didn't do this it was akin to always using the target x86-64 no matter what vector instruction set exists on the host target. Sniffing the CPU feature set is much simpler than sniffing the host cuda feature set, unfortunately.

**Comment by steven-johnson** [comment]
I get all of that, but I think that invoking the JIT by way of a Target() ctor is extremely suprising -- Target is meant to be a pretty lightweight class, this is not a side effect I'd expect. Perhaps a more explicit API to request this behavior would be appropriate.

(BTW: I only even noticed this because certain arm32-under-qemu generators are crashing somewhere inside this point... but the arm32 jit (in general) seems to work fine under qemu, so something else here is odd.)

**Comment by abadams** [comment]
Yeah, I had trouble thinking of any other way to do it. We resolve "host" eagerly in the target constructor, and that requires some sniffing.

We could defer that somehow. We could make the sniffing cheaper.

Honestly it's weird that we JIT-compile the host runtimes at all. They could be statically compiled and linked into libHalide and things would be faster. Would clean up a bunch of symbol resolution nonsense when jitting. It's just annoying to set up the build for that.

**Comment by steven-johnson** [comment]
The eager host resolution is pretty cheap though - lazily sniff in `calculate_host_target()` and cache it, and the sniffing itself is arguably not much more expensive than the string ops we do elsewhere.

Agreed re: jit-compiling the host runtimes. Is it really hard to set up the build for that? (Out of curiosity, was it attempted long ago and abandoned?)

**Comment by abadams** [comment]
The *cpu* eager host resolution is cheap. I'd argue that this is also just host resolution and should be done at the same time as CPU features.

We used to have name collisions between different runtime modules, because they provided alternative implementations of things, so we never tried to link them all at once. I'm not sure if it's still the case that it ever makes sense to simultaneously link two modules that provide the same names. (E.g. posix_threads and posix_threads_tsan) - that's the thing that might foil just prelinking all of the runtime modules for the host system into libHalide, unless we can cleverly namespace the symbols somehow.

**Comment by abadams** [comment]
An example of something people might reasonably expect to work in the same way on CPU and GPU:

```
Target t(some string that happens to be "host-cuda");
 ...
if (t.has_feature(Target::AVX512)) {
 ...
}
if (t.has_feature(Target::CUDACapability50) {
 ...
}
```

**Comment by steven-johnson** [comment]
> collisions between different runtime modules

Ack, our existing runtime design with weak-symbol-fu strikes again.

If I had an arbitrary amount of time to work on something Halide, I want Halide Runtime 2.0, in which we replace all the weak-symbol stuff *and* user-context with a mandatory 'runtime-context' which contains a table of function ptrs for the runtime code (as well as hooks for what people are using user-context for now). presto, no name collision issues, and you can even mix and match different runtime combinations within the same object file if you like. (The trick would be figuring out if it's possible/practical to do this without being a breaking change...)

**Comment by steven-johnson** [comment]
Back on the main topic though -- I'd argue that the sniffing for "host-cuda" is really only appropriate when the eventual goal is to JIT in the first place (since we know that the 'host' is where it's gonna be run).

For AOT compilation, 'host' may or may not have any relation to where we'll execute (and thus the capabilities of cuda there); indeed, inside Google, the AOT-compilation 'host' machine may not even be in the same time zone as the intended execution environment.

**Comment by abadams** [comment]
"'host' may or may not have any relation to where we'll execute"

That seems really weird to me.

**Comment by abadams** [comment]
To elaborate: Why doesn't the same objection apply to AVX512 sniffing? You're in a situation where you know the build machine has the same CPU as the target machine, but other characteristics might be different?

**Comment by steven-johnson** [comment]
Oh, the same objection does apply -- what I realized after I hit "enter" is that in the Google case, we never use "host" (we always use an explicit target string). It's really just Halide's own Makefile/CMake setup that takes this shortcut.

**Comment by abadams** [comment]
If you're about to JIT, clearly we need to initialize the JIT module anyway. So the case where this does something weird is if you're AOT compiling with the target host-cuda, and it surprisingly does a JIT compilation prior to the intended compilation just to figure out what "host" means.

So I think we just want to make this cheaper somehow. We can't turn it off without a major performance regression for most people using cuda (and reintroducing a large foot-gun).

The options I can think of to get the cuda capability are loading libcuda redundantly outside of the runtime, shelling out to a tool (e.g. to nvidia-smi), statically linking the runtime in (a major endeavor), caching the host cuda capability somewhere persistent...

But this really is just a performance optimization. The current implementation of the feature was the simplest thing that fixed the foot-gun.
